### PR TITLE
Authentication Setup with NextAuth

### PR DESCRIPTION
**Pull request by hellodanielwilliams**
## TL/DR

- Sets up nextauth with Credentials Provider
- Implements password hashing with bcrypt
- Adds register page for new users
- Adds log in page for existing user
- Adds profile page that can be accessed by logged in user
- Wraps a session provider around app

### Tailwind

- Needed to remake the `globals.css` file to access the tailwind classes for basic styling. Now importing globals in `layout.tsx`.
- Styling is very simple now, can be changed to fit with rest of site later

### `.env`
- Added `NEXTAUTH_URL=http://localhost:3000/`
- Added `NEXTAUTH_SECRET` - will message the secret key in group if this gets approved and merged (that’s if .env is actually being ignored correctly… 🤔)

### Packages

- `Next-auth` - used throughout for handling login, session, sign out  etc. configured in `/api/auth/[…nextauth]/route.ts`
- `bcrypt.js` - used  to handle password hashing and matching

### `Providers.tsx`

- The app needs to be wrapped in a `SessionProvider`, but this can’t be done directly in `layout.tsx` as it needs to `“use client”`
- `AuthProvider` is setup with `SessionProvider` as export in `Providers.tsx`, then wrapped around children in `layout.tsx`

## New Pages

The functionality of each of the pages has been extracted out into a new `/components directory` - this allows the components to `“use client”` 

### Login

- Email and password fields with `useState` to control state
- On submit, tries to `signin` (function from `next-auth`) using credentials provided
- If signin fails, form resets,  error message displays
- If successful, routes to `/profile` using `useRouter`, which comes from `next/navigation`, NOT next/router

### Register
- `useState` to control input fields again
- On submit, first checks fields are not empty, displays error if so
- Then fires a `POST` request to `/api/userExists` which searches users by inputted email, displays error if match found
- If no match, sends `POST` to `api/register`, which hashes password using bcrypt and  calls `insertUser` from models to insert user into db

### Profile
- Uses `useSession` to check if a user is logged in and to get the data
- Redirects to /login if no session
- Displays user name and email from the session
- `signOut` function (from `next-auth`) handles sign out, redirects to login

## Typescript Notes
- `e: React.FormEvent<HTMLFormElement>` seems to make it happy for an onSubmit event
- `const form = e.target as HTMLFormElement` works for a form
- `Promise<User | null>` works for the authorize callback
- There’s a `NextAPIHandler` type

## Issues
- Getting a nextJs error in the browser that gives a warning about hydration, and it not expecting to find a div in div.
- Not sure why this is happening but it does not stop the pages from rendering.

### CODE REVIEW

**Review by elliejaneeeeee — APPROVED** [review]
Amazing!!! I loved the TLDR message it made it really easy to see how it worked and what changes were made, I really appreciate it! This looks incredible thank you so much for working on it out of hours :D You made a massive step for us in the project
